backend/app/core/search.py

```diff
@@ -0,0 +1,466 @@
+import time
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Tuple
+
+from opensearchpy import OpenSearch
+from sentence_transformers import SentenceTransformer
+
+from app.db.schemas.search import (
+    OpenSearchResponseDescriptionMatch,
+    OpenSearchResponseNameMatch,
+    OpenSearchResponseMatchBase,
+    OpenSearchResponsePassageMatch,
+    SearchRequestBody,
+    SearchResponseBody,
+    SearchResponseDocument,
+    SearchResponseDocumentPassage,
+)
+
+_DEFAULT_INNER_PRODUCT_THRESHOLD: float = 70.0
+_DEFAULT_MAX_DOC_COUNT: int = 100
+_DEFAULT_MAX_PASSAGES_PER_DOC: int = 10
+_DEFAULT_N_PASSAGES_TO_SAMPLE_PER_SHARD: int = 5000
+_DEFAULT_NAME_BOOST: int = 100
+_DEFAULT_DESCRIPTION_BOOST: int = 100
+_DEFAULT_OPENSEARCH_INDEX = "navigator"
+_DEFAULT_OPENSEARCH_REQUEST_TIMEOUT = 30
+_DEFAULT_OPENSEARCH_PREFERENCE = "prototype_user"  # TODO: document what this means
+_DEFAULT_OPENSEARCH_EMBEDDING_DIM = 768
+
+# TODO: Allow this to be configured
+_ENCODER = SentenceTransformer(model_name_or_path="msmarco-distilbert-dot-v5")
+
+_OPENSEARCH_NAME_KEY = "for_search_action_name"
+_OPENSEARCH_DESCRIPTION_KEY = "for_search_action_description"
+_OPENSEARCH_DESCRIPTION_EMBEDDING_KEY = "action_description_embedding"
+_OPENSEARCH_INDEX_KEY = "action_name_and_id"
+_OPENSEARCH_TEXT_BLOCK_KEY = "text_block_id"
+
+
+@dataclass
+class OpenSearchConfig:
+    """Config for accessing an OpenSearch instance."""
+
+    url: str
+    username: str
+    password: str
+    index_name: str = _DEFAULT_OPENSEARCH_INDEX
+    embedding_dim: int = _DEFAULT_OPENSEARCH_EMBEDDING_DIM
+    request_timeout: int = _DEFAULT_OPENSEARCH_REQUEST_TIMEOUT
+    preference: str = _DEFAULT_OPENSEARCH_PREFERENCE
+    use_ssl: bool = False
+    verify_certs: bool = False
+    ssl_show_warnings: bool = False
```

Comment:
The values of these parameters will probably change when we connect to the managed service as it uses SSL. I guess we want them as environment variables so we can still run tests against unsecured Opensearch instances in dev/test?